pyadjoint/optimization/rol_solver_checkpointed.py

```diff
@@ -0,0 +1,283 @@
+                return None
+
+            subdir_name = os.path.join(
+                self.__ROL_checkpoint_dir__,
+                f"iteration_{self.__index__ - 1}")
+
+            if os.path.isdir(subdir_name):
+                return subdir_name
+            else:
+                return None
+
+    ROLCheckpointManager = __ROLCheckpointManager__()
+
+    class CheckpointedROLVector(ROLVector):
+        def __init__(self, dat, inner_product="L2"):
+            super().__init__(dat, inner_product)
+
+            with CheckpointFile(ROLCheckpointManager.get_mesh_name()[0],
+                                mode='r') as fi:
+                self.mesh = fi.load_mesh(ROLCheckpointManager.get_mesh_name()[1])
+
+        def load(self):
+            """Load our data once self.dat is populated"""
+            with CheckpointFile(self.fname, mode='r') as ckpnt:
+                for i, _ in enumerate(self.dat):
+                    self.dat[i] = \
+                        ckpnt.load_function(self.mesh, name=f'dat_{i}')
+
+        def save(self, fname):
+            with CheckpointFile(fname, mode='w') as ckpnt:
+                ckpnt.save_mesh(self.mesh)
+                for i, f in enumerate(self.dat):
+                    ckpnt.save_function(f, name=f"dat_{i}")
+
+        def clone(self):
+            dat = []
+            for x in self.dat:
+                dat.append(x._ad_copy())
+            res = CheckpointedROLVector(dat, inner_product=self.inner_product)
+            res.scale(0.0)
+            return res
+
+        def __setstate__(self, state):
+            """Set the state from unpickling
+
+            Requires self.dat to be separately set, then self.load()
+            can be called.
+            """
+
+            # initialise C++ state
+            super().__init__(state)
+
+            self.fname, self.inner_product = state
+
+            with CheckpointFile(ROLCheckpointManager.get_mesh_name()[0],
+                                mode='r') as fi:
+                self.mesh = fi.load_mesh(ROLCheckpointManager.get_mesh_name()[1])
+
+            vector_registry.append(self)
+
+        def __getstate__(self):
+            """Return a state tuple suitable for pickling"""
+
+            fname = os.path.join(
+                ROLCheckpointManager.get_checkpoint_dir(),
+                "vector_checkpoint_{}.h5".format(utils._new_uid()))
+            self.save(fname)
+
+            return (fname, self.inner_product)
+
+    class CheckPointedROLSolver(ROLSolver):
+        def __init__(self, problem, parameters, inner_product="L2"):
+            super().__init__(problem, parameters)
+            OptimizationSolver.__init__(self, problem, parameters)
+            self.rolobjective = ROLObjective(problem.reduced_functional)
+            x = [p.tape_value() for p in self.problem.reduced_functional.controls]
+            self.rolvector = CheckpointedROLVector(x, inner_product=inner_product)
+            self.params_dict = parameters
```

Comment:
I wonder if instead, we could create an instance of `_ROLCheckpointManager` here, and access it through the ROL solver rather than a global singleton. Then `CheckpointedROLVector` would need to know about the specific manager instance (could do it as a factory or just another parameter).